# Курс "Python для анализа данных"

---
#4.1.4 Иерархическое индексирование

## Где мы сейчас?


<html>
 <head>
  <meta charset="utf-8">
 </head>
 <body>
  <ul>
    <li>1. Введение в анализ данных и разработку на языке Python </li> <!-2.1->
    <li>2. Основы языка </li> <!-2.1->
    <li>3. Библиотеки NumPy и SciPy для решения задач анализа данных и машинного обучения </li> <!-2.1->
    <li>4. <strong>Библиотека Pandas для работы с высокоуровневыми структурами данных</strong> <i><- Вот в этой главе!</i>
     <ul>
      <li><strong>4.1 Введение в структуры данных Pandas</strong> <i><- Вот в этом параграфе!</i></li>
       <ul>
        <li>4.1.1 Работа с объектами Series, DataFrame. Переиндексация, доступ по индексу, выборка, фильтрация</li>
        <li>4.1.2 Применение функций и отображений, арифметические операции и выравнивание данных</li>
        <li>4.1.3 Редукция и вычисление описательных статистик. Обработка отсутствующих данных</li>
        <li><strong>4.1.4 Иерархическое индексирование</strong><i> <- Вот в этом пункте!</i></li>
       </ul>
      <li>4.2 Переформатирование данных: очистка, преобразование, слияние, изменение формы</li>
     </ul>
    </li>   
  </ul>
 </body>
</html>

## О чем будем говорить?


О том, как можно организовывать несколько уровней индексирования по одной оси в объектах ```Series``` и ```DataFrame```. А также более подробно посмотрим на индексные объекты. 

## Ключевые слова, понятия и выражения






*   ```pandas```
*   ```groupby```
*   ```MultiIndex```

## Материал

### Индексные объекты в ```pandas```

![alt text](https://drive.google.com/uc?id=16AmI0-pVXLq0OvPNt53-Fvgj-enA01ro)

In [ ]:
import pandas as pd
import numpy as np

print('Пример с Index')
df = pd.DataFrame(np.arange(12).reshape(3,4), index=['a','b','c'])
print(df.index)
df.head()

Пример с Index
Index(['a', 'b', 'c'], dtype='object')


0  1   2   3
a  0  1   2   3
b  4  5   6   7
c  8  9  10  11

In [ ]:
print('Пример с Int64Index')
df = pd.DataFrame(np.arange(12).reshape(3,4), index=[1,2,3])
print(df.index)
df.head()

Пример с Int64Index
Int64Index([1, 2, 3], dtype='int64')


0  1   2   3
1  0  1   2   3
2  4  5   6   7
3  8  9  10  11

In [ ]:
from google.colab import files
files.upload()

Saving Occupancy.csv to Occupancy (1).csv
Saving Wine.csv to Wine (1).csv


{'Occupancy.csv': b'"date","Temperature","Humidity","Light","CO2","HumidityRatio","Occupancy"\n"1","2015-02-04 17:51:00",23.18,27.272,426,721.25,0.00479298817650529,1\n"2","2015-02-04 17:51:59",23.15,27.2675,429.5,714,0.00478344094931065,1\n"3","2015-02-04 17:53:00",23.15,27.245,426,713.5,0.00477946352442199,1\n"4","2015-02-04 17:54:00",23.15,27.2,426,708.25,0.00477150882608175,1\n"5","2015-02-04 17:55:00",23.1,27.2,426,704.5,0.00475699293331518,1\n"6","2015-02-04 17:55:59",23.1,27.2,419,701,0.00475699293331518,1\n"7","2015-02-04 17:57:00",23.1,27.2,419,701.666666666667,0.00475699293331518,1\n"8","2015-02-04 17:57:59",23.1,27.2,419,699,0.00475699293331518,1\n"9","2015-02-04 17:58:59",23.1,27.2,419,689.333333333333,0.00475699293331518,1\n"10","2015-02-04 18:00:00",23.075,27.175,419,688,0.00474535071966655,1\n"11","2015-02-04 18:01:00",23.075,27.15,419,690.25,0.00474095189694268,1\n"12","2015-02-04 18:02:00",23.1,27.1,419,691,0.00473937073052061,1\n"13","2015-02-04 18:03:00",23.1,27.1666

In [ ]:
print('Пример с DatetimeIndex')
data_occupancy = pd.read_csv('Occupancy.csv', sep=',')
print(data_occupancy.index)
data_occupancy.head()

Пример с DatetimeIndex
Int64Index([   1,    2,    3,    4,    5,    6,    7,    8,    9,   10,
            ...
            8134, 8135, 8136, 8137, 8138, 8139, 8140, 8141, 8142, 8143],
           dtype='int64', length=8143)


date  Temperature  Humidity  ...     CO2  HumidityRatio  Occupancy
1  2015-02-04 17:51:00        23.18   27.2720  ...  721.25       0.004793          1
2  2015-02-04 17:51:59        23.15   27.2675  ...  714.00       0.004783          1
3  2015-02-04 17:53:00        23.15   27.2450  ...  713.50       0.004779          1
4  2015-02-04 17:54:00        23.15   27.2000  ...  708.25       0.004772          1
5  2015-02-04 17:55:00        23.10   27.2000  ...  704.50       0.004757          1

[5 rows x 7 columns]

In [ ]:
data_occupancy.index = pd.to_datetime(data_occupancy.date)
data_occupancy.drop('date', axis=1, inplace=True)
print(data_occupancy.index)
data_occupancy.head()

DatetimeIndex(['2015-02-04 17:51:00', '2015-02-04 17:51:59',
               '2015-02-04 17:53:00', '2015-02-04 17:54:00',
               '2015-02-04 17:55:00', '2015-02-04 17:55:59',
               '2015-02-04 17:57:00', '2015-02-04 17:57:59',
               '2015-02-04 17:58:59', '2015-02-04 18:00:00',
               ...
               '2015-02-10 09:23:59', '2015-02-10 09:24:59',
               '2015-02-10 09:26:00', '2015-02-10 09:27:00',
               '2015-02-10 09:28:00', '2015-02-10 09:29:00',
               '2015-02-10 09:29:59', '2015-02-10 09:30:59',
               '2015-02-10 09:32:00', '2015-02-10 09:33:00'],
              dtype='datetime64[ns]', name='date', length=8143, freq=None)


Temperature  Humidity  ...  HumidityRatio  Occupancy
date                                        ...                          
2015-02-04 17:51:00        23.18   27.2720  ...       0.004793          1
2015-02-04 17:51:59        23.15   27.2675  ...       0.004783          1
2015-02-04 17:53:00        23.15   27.2450  ...       0.004779          1
2015-02-04 17:54:00        23.15   27.2000  ...       0.004772          1
2015-02-04 17:55:00        23.10   27.2000  ...       0.004757          1

[5 rows x 6 columns]

In [ ]:
data_occupancy['2015-02-04 17:51:00' : '2015-02-04 17:55:00']

Temperature  Humidity  ...  HumidityRatio  Occupancy
date                                        ...                          
2015-02-04 17:51:00        23.18   27.2720  ...       0.004793          1
2015-02-04 17:51:59        23.15   27.2675  ...       0.004783          1
2015-02-04 17:53:00        23.15   27.2450  ...       0.004779          1
2015-02-04 17:54:00        23.15   27.2000  ...       0.004772          1
2015-02-04 17:55:00        23.10   27.2000  ...       0.004757          1

[5 rows x 6 columns]

In [ ]:
data_occupancy[data_occupancy.index.day == 6].head()

Temperature  Humidity  ...  HumidityRatio  Occupancy
date                                        ...                          
2015-02-06 00:00:00         20.2     21.29  ...       0.003110          0
2015-02-06 00:01:00         20.2     21.20  ...       0.003097          0
2015-02-06 00:02:00         20.2     21.29  ...       0.003110          0
2015-02-06 00:03:00         20.2     21.29  ...       0.003110          0
2015-02-06 00:04:00         20.2     21.29  ...       0.003110          0

[5 rows x 6 columns]

In [ ]:
data_occupancy['2015-02-08'].head()

Temperature   Humidity  ...  HumidityRatio  Occupancy
date                                         ...                          
2015-02-08 00:00:00         19.6  29.390000  ...       0.004143          0
2015-02-08 00:01:00         19.6  29.463333  ...       0.004154          0
2015-02-08 00:02:00         19.6  29.500000  ...       0.004159          0
2015-02-08 00:03:00         19.6  29.445000  ...       0.004151          0
2015-02-08 00:04:00         19.6  29.500000  ...       0.004159          0

[5 rows x 6 columns]

In [ ]:
data_occupancy.resample('20min').sum()

Temperature    Humidity  ...  HumidityRatio  Occupancy
date                                          ...                          
2015-02-04 17:40:00   208.130000  244.984500  ...       0.042912          9
2015-02-04 18:00:00   458.790000  546.431667  ...       0.094636          7
2015-02-04 18:20:00   451.901667  548.818333  ...       0.093077          0
2015-02-04 18:40:00   446.900000  548.046667  ...       0.091531          0
2015-02-04 19:00:00   444.127500  545.430000  ...       0.090320          0
...                          ...         ...  ...            ...        ...
2015-02-10 08:00:00   405.950000  661.910000  ...       0.097551          0
2015-02-10 08:20:00   405.916667  664.106667  ...       0.097867          2
2015-02-10 08:40:00   389.041667  639.821667  ...       0.095365         18
2015-02-10 09:00:00   415.539167  695.019167  ...       0.105584         20
2015-02-10 09:20:00   294.433750  502.851250  ...       0.077620         14

[408 rows x 6 columns]

### Использование MultiIndex

In [ ]:
data = pd.Series(np.random.randn(10),
                 index= [['a','a','a','b','b','b','c','c','d','d'],
                         [1,2,3,1,2,3,1,2,2,3,]])
print(data.index)
data

MultiIndex([('a', 1),
            ('a', 2),
            ('a', 3),
            ('b', 1),
            ('b', 2),
            ('b', 3),
            ('c', 1),
            ('c', 2),
            ('d', 2),
            ('d', 3)],
           )


a  1   -0.099308
   2    0.524431
   3    0.274392
b  1    2.032126
   2   -0.809222
   3    0.730071
c  1    0.772442
   2   -0.350429
d  2   -1.033105
   3   -0.974617
dtype: float64

In [ ]:
print('Обращение по частичному индексу:')
data['a']

Обращение по частичному индексу:


1   -0.099308
2    0.524431
3    0.274392
dtype: float64

In [ ]:
data[:,2]

a    0.524431
b   -0.809222
c   -0.350429
d   -1.033105
dtype: float64

In [ ]:
data['b':'c']

b  1    2.032126
   2   -0.809222
   3    0.730071
c  1    0.772442
   2   -0.350429
dtype: float64

### Использование ```GroupBy``` для получения иерархии индексов в ```DataFrame```

In [ ]:
data_wine = pd.read_csv('Wine.csv', sep=',',index_col=0)
print(data_wine.index)
data_wine.head()

Int64Index([   0,    1,    2,    3,    4,    5,    6,    7,    8,    9,
            ...
            4989, 4990, 4991, 4992, 4993, 4994, 4995, 4996, 4997, 4998],
           dtype='int64', length=4999)


country  ...                   winery\t\t\t
0      US  ...                    Heitz\t\t\t
1   Spain  ...  Bodega Carmen Rodríguez\t\t\t
2      US  ...                 Macauley\t\t\t
3      US  ...                    Ponzi\t\t\t
4  France  ...     Domaine de la Bégude\t\t\t

[5 rows x 10 columns]

In [ ]:
data_wine.groupby('country').count()

description  designation  points  ...  region_2  variety  winery\t\t\t
country                                         ...                                 
Albania                 2            0       2  ...         0        2             2
Argentina             114           93     114  ...         0      114           114
Australia              33           28      33  ...         0       33            33
Austria                74           53      74  ...         0       74            74
Bulgaria               15           13      15  ...         0       15            15
Canada                 16           13      16  ...         0       16            16
Chile                  88           80      88  ...         0       88            88
Croatia                 6            5       6  ...         0        6             6
France                908          564     908  ...         0      908           908
Germany               165          158     165  ...         0      165           165
Greece                 19           11      19  ...         0       19            19
Hungary                11            5      11  ...         0       11            11
India                   5            2       5  ...         0        5             5
Israel                 29           26      29  ...         0       29            29
Italy                 737          526     737  ...         0      737           737
Lebanon                 5            4       5  ...         0        5             5
Macedonia               2            1       2  ...         0        2             2
Moldova                 8            8       8  ...         0        8             8
Morocco                 7            3       7  ...         0        7             7
New Zealand            24           18      24  ...         0       24            24
Portugal              229          205     229  ...         0      229           229
Romania                 5            5       5  ...         0        5             5
Serbia                  6            5       6  ...         0        6             6
Slovenia               14           12      14  ...         0       14            14
South Africa           16            9      16  ...         0       16            16
Spain                 294          239     294  ...         0      294           294
Switzerland             2            1       2  ...         0        2             2
Turkey                  9            7       9  ...         0        9             9
US                   2152         1538    2152  ...      2051     2152          2152
Uruguay                 2            2       2  ...         0        2             2

[30 rows x 9 columns]

In [ ]:
data_wine[data_wine['country'] == 'Macedonia']

In [ ]:
data_wine_c_p = data_wine.groupby(['country', 'province']).max()
print(data_wine_c_p.index)
data_wine_c_p

MultiIndex([(  'Albania',          'Mirditë'),
            ('Argentina', 'Mendoza Province'),
            ('Argentina',            'Other'),
            ('Australia',  'Australia Other'),
            ('Australia',  'South Australia'),
            ('Australia',         'Tasmania'),
            ('Australia',         'Victoria'),
            (  'Austria',       'Burgenland'),
            (  'Austria',        'Carnuntum'),
            (  'Austria',        'Eisenberg'),
            ...
            (       'US',       'California'),
            (       'US',         'Colorado'),
            (       'US',            'Idaho'),
            (       'US',         'New York'),
            (       'US',   'North Carolina'),
            (       'US',           'Oregon'),
            (       'US',         'Virginia'),
            (       'US',       'Washington'),
            (  'Uruguay',          'Juanico'),
            (  'Uruguay',       'Montevideo')],
           names=['country', 'province'], l

description  ...           winery\t\t\t
country   province                                                             ...                       
Albania   Mirditë           This garnet-colored wine made from 100% Kallme...  ...           Arbëri\t\t\t
Argentina Mendoza Province  Unconvincing aromas are hard to describe. This...  ...         Zuccardi\t\t\t
          Other             Wiry herbal aromas of blueberry and currant le...  ...            Verum\t\t\t
Australia Australia Other   A terrific value in sparkling wine, this sligh...  ...         Taltarni\t\t\t
          South Australia   With only 12% new oak and 28% of the fruit com...  ...          Yalumba\t\t\t
...                                                                       ...  ...                    ...
US        Oregon            Young vine flavors of bright raspberry fruit s...  ...        Winderlea\t\t\t
          Virginia          White apricot, spice and lemon flavors give ch...  ...          Veramar\t\t\t
          Washington        Wood takes center stage on this wine. Char, da...  ...  Woodward Canyon\t\t\t
Uruguay   Juanico           This mature Bordeaux-style blend is earthy on ...  ...   Familia Deicas\t\t\t
          Montevideo        Bouza ranks as one of Uruguay's top wineries, ...  ...            Bouza\t\t\t

[166 rows x 5 columns]

In [ ]:
data_wine_c_p.loc['France']

## Дополнительные материалы и литература



*   Уэс Маккинли, Python и анализ данных/ Пер. с англ. Слинкин А. А. - М.: ДМК Пресс, 2015. - 482 с.: ил. С. 166